src/utils/io/progress_log.rs

```diff
@@ -0,0 +1,234 @@
+// Copyright 2019-2023 ChainSafe Systems
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use humantime::format_duration;
+use std::time::{Duration, Instant};
+
+//use std::borrow::Cow;
+//use std::convert::TryFrom;
+use std::io::{self};
+//use std::io::IoSliceMut;
+//use std::iter::FusedIterator;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use tokio::io::{ReadBuf, SeekFrom};
+
+use log::info;
+
+const UPDATE_FREQUENCY: Duration = Duration::from_millis(500);
+
+pub fn wrap_iter<Inner>(
+    message: &str,
+    into_iter: impl IntoIterator<IntoIter = Inner>,
+) -> WithProgressIter<Inner>
+where
+    Inner: Iterator,
+{
+    let inner = into_iter.into_iter();
+    let (lower_bound, upper_bound) = inner.size_hint();
+    let total_items = upper_bound.unwrap_or(lower_bound) as u64;
+    WithProgressIter {
+        inner,
+        progress: WithProgress::new(message, total_items),
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct WithProgressIter<Inner> {
+    inner: Inner,
+    progress: WithProgress,
+}
+
+impl<Inner> Iterator for WithProgressIter<Inner>
+where
+    Inner: Iterator,
+{
+    type Item = Inner::Item;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.inner.next() {
+            Some(item) => {
+                self.progress.inc(1);
+                Some(item)
+            }
+            None => {
+                // TODO handle fusing
+                println!("finished {} items", self.progress.completed_items);
+                None
+            }
+        }
+    }
+}
+
+pub fn wrap_stream<S: futures_core::Stream>(message: &str, stream: S) -> WithProgressStream<S> {
+    let (lower_bound, upper_bound) = stream.size_hint();
+    let total_items = upper_bound.unwrap_or(lower_bound) as u64;
```

Comment:
I'm not sure this will work well. We might have to specify the upper bound separately.

Reply:
No, I don't think we should use the size hint at all.

Reply:
> No, I don't think we should use the size hint at all.

Yeah, I think using inaccurate values for progress tracking is only going to confuse people and also won't reflect the actual state of things.